pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,496 @@
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	trafficControlinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	trafficControllisters "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+	"antrea.io/antrea/pkg/ovs/ovsconfig"
+)
+
+const (
+	controllerName = "TrafficControlController"
+	defaultWorkers = 4
+	resyncPeriod   = 0
+)
+
+type tcState struct {
+	ofPorts      sets.Int32
+	pods         sets.String
+	dstOfInPort  uint32
+	dstOfOutPort uint32
+}
+
+type Controller struct {
+	nodeName string
+
+	ofClient        openflow.Client
+	ovsBridgeClient ovsconfig.OVSBridgeClient
+
+	interfaceStore interfacestore.InterfaceStore
+
+	podInformer cache.SharedIndexInformer
+	podLister   corelisters.PodLister
+
+	namespaceLister corelisters.NamespaceLister
+
+	tcStates      map[string]*tcState
+	tcStatesMutex sync.RWMutex
+
+	ovsPorts map[string]ovsconfig.OVSPortData
+
+	trafficControlInformer cache.SharedIndexInformer
+	// trafficControlLister
+	trafficControlLister trafficControllisters.TrafficControlLister
+	// trafficControlListerSynced is a function which returns true if the TrafficControl shared informer has been synced at least once.
+	trafficControlListerSynced cache.InformerSynced
+
+	// queue maintains the TrafficControlGroup objects that need to be synced.
+	queue workqueue.RateLimitingInterface
+}
+
+func NewTrafficControlController(nodeName string, ofClient openflow.Client,
+	interfaceStore interfacestore.InterfaceStore, ovsBridgeClient ovsconfig.OVSBridgeClient,
+	tcInformer trafficControlinformers.TrafficControlInformer, podInformer cache.SharedIndexInformer, namespaceInformer v12.NamespaceInformer) *Controller {
+	c := &Controller{
+		nodeName:                   nodeName,
+		ofClient:                   ofClient,
+		ovsBridgeClient:            ovsBridgeClient,
+		interfaceStore:             interfaceStore,
+		trafficControlInformer:     tcInformer.Informer(),
+		trafficControlLister:       tcInformer.Lister(),
+		trafficControlListerSynced: tcInformer.Informer().HasSynced,
+		podInformer:                podInformer,
+		podLister:                  corelisters.NewPodLister(podInformer.GetIndexer()),
+		namespaceLister:            namespaceInformer.Lister(),
+		tcStates:                   map[string]*tcState{},
+		queue:                      workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(5*time.Second, 300*time.Second), "trafficControl"),
+	}
+	c.trafficControlInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addTC,
+			UpdateFunc: c.updateTC,
+			DeleteFunc: c.deleteTC,
+		},
+		resyncPeriod,
+	)
+	return c
+}
+
+func (c *Controller) addTC(obj interface{}) {
+	tc := obj.(*v1alpha2.TrafficControl)
+	klog.InfoS("Processing TrafficControl ADD event", "trafficControl", tc.Name, "podSelector", tc.Spec.AppliedTo)
```

Comment:
```suggestion
	klog.InfoS("Processing TrafficControl ADD event", "trafficControl", tc.Name, "appliedTo", tc.Spec.AppliedTo)
```
as it may have namespaceSelector